tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+    pub fn set_comb(&mut self, is_comb: bool) {
+        self.is_comb = is_comb;
+    }
+
+    /// Adds an input port `name` to this component.
+    pub fn add_input<S: ToString>(&mut self, name: S, width: usize) {
+        self.add_port(Port::new_in_comp(name, width), true);
+    }
+
+    /// Adds an output port `name` to this component.
+    pub fn add_output<S: ToString>(&mut self, name: S, width: usize) {
+        self.add_port(Port::new_in_comp(name, width), false);
+    }
+
+    /// [`Component::add_input`] or [`Component::add_output`] may be more
+    /// useful.
+    pub fn add_port(&mut self, port: Port, is_input: bool) {
+        if is_input {
+            self.inputs.push(port);
+        } else {
+            self.outputs.push(port);
+        }
+    }
+
+    /// Constructs a new cell named `name` that instatiates `inst` with
+    /// arguments `args`.
+    pub fn cell<S: ToString, T: ToString>(
+        &mut self,
+        is_ref: bool,
+        name: S,
+        inst: T,
+        args: Vec<u64>,
+    ) -> RRC<Cell> {
+        let cell = rrc(Cell {
+            is_ref,
+            attributes: vec![],
+            name: name.to_string(),
+            inst: inst.to_string(),
+            args,
+        });
+        self.cells.push(cell.clone());
+        cell
+    }
+
+    /// Establishes a continuous assignment guarded by `guard` between `lhs` and
+    /// `rhs`.
+    pub fn assign(&mut self, lhs: Port, rhs: Port, guard: Guard) {
+        self.continuous_assignments
+            .push(Assignment::new(lhs, rhs, guard));
+    }
+
+    /// Use [`declare_group!`] and [`build_group!`] instead.
+    pub fn group<S: ToString>(&mut self, name: S) -> RRC<Group> {
+        let group = rrc(Group::new(name, false));
+        self.groups.push(group.clone());
+        group
+    }
+
+    /// Use [`declare_group!`] and [`build_group!`] instead.
+    pub fn comb_group<S: ToString>(&mut self, name: S) -> RRC<Group> {
+        let group = rrc(Group::new(name, true));
+        self.groups.push(group.clone());
+        group
+    }
+
+    /// Sets the root control node for this component. Use [`build_control!`] to
+    /// construct this node.
+    pub fn set_control(&mut self, control: Control) {
+        self.control = control;
+    }
+
+    /// Opens an indented section within a component and runs a callback after
+    /// applying another indent. For instance, when called with a `name` of
+    /// "cells", this function allows formatting of:
+    /// ```
+    /// ...
+    ///   cells {
+    ///     ...
+    ///   }
+    /// ...
+    /// ```
+    fn brace_section<S: ToString, F>(
+        &self,
+        f: &mut IndentFormatter<'_, '_>,
+        name: S,
+        body: F,
+    ) -> fmt::Result
+    where
+        F: FnOnce(&mut IndentFormatter<'_, '_>) -> fmt::Result,
+    {
+        f.push();
+        writeln!(f, "{} {{", name.to_string(),)?;
+        f.push();
+        let result = body(f);
```

Comment:
I would just throw this like normal and return `Ok(())` at the end, rather than waiting